include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -306,10 +306,105 @@ MultigridPreconditionerBase<dim, Number>::check_levels(std::vector<MGLevelInfo>
   }
 }
 
+/**
+ * Helper function for creating a geometric coarsening sequence: Replicate a
+ * parallel::distributed::Triangulation completely in a serial
+ * triangulation. This can potentially be very memory-consuming, so the
+ * triangulation is only filled on the first rank of a compute node, i.e.,
+ * those MPI ranks where `rank % n_processes_per_node` is zero.
+ */
+template<int dim, int spacedim>
+void
+replicate_distributed_triangulation_on_node(
+  parallel::distributed::Triangulation<dim, spacedim> const & distributed_tria,
+  MPI_Comm const &                                            mpi_comm,
+  unsigned int const                                          n_processes_per_node,
+  dealii::Triangulation<dim, spacedim> &                      serial_tria)
+{
+  unsigned int const my_rank                  = Utilities::MPI::this_mpi_process(mpi_comm);
+  bool const         is_first_process_on_node = (my_rank % n_processes_per_node) == 0;
```

Comment:
The same code is used somewhere below to identify the first process on node. For me, this code is so general that it should become a utility function in ExaDG -> `is_first_process_on_node(mpi_comm, n_processes_per_node)`. Ideally, `n_processes_per_node` would not be part of the interface but happen inside that function, but this design decision probably depends on how expensive it is to call the lambda `n_processes_per_node`.

Regarding this function, I believe that passing `is_first_process_on_node` would be a better design. I don't see what `n_processes_per_node` is used for apart from determining `is_first_process_on_node`. So my preferred option in terms of design would be a call like `is_first_process_on_node(mpi_comm)` outside of this function, and passing the result to this function. Does this make sense?

Reply:
Good idea. I added the function in `exadg/utilities/mpi.h` as `identify_first_process_on_node` - I want to return both the question of whether it is the first process, and the number of processes per node that we also need in this function, as it would be expensive to recompute this. (If I measured correctly, we're talking about more than 10 seconds per call on 3k nodes.)

Reply:
Have a look at the new commit 02b166b6 that addresses this issue.